diff -r fc7b9e4d2d03 libpurple/media.c
--- a/libpurple/media.c	Tue Nov 26 00:30:54 2019 -0500
+++ b/libpurple/media.c	Thu Dec 12 09:33:37 2019 -0500
@@ -600,7 +600,7 @@
 		g_signal_emit(media, purple_media_signals[STATE_CHANGED],
 				0, PURPLE_MEDIA_STATE_END,
 				stream->session->id, stream->participant);
-
+printf("free session 1\n");
 		media->priv->streams =
 				g_list_remove(media->priv->streams, stream);
 
@@ -631,6 +631,7 @@
 		g_signal_emit(media, purple_media_signals[STATE_CHANGED],
 				0, PURPLE_MEDIA_STATE_END,
 				session->id, NULL);
+printf("free session 2\n");
 
 		g_hash_table_remove(media->priv->sessions, session->id);
 		purple_media_session_free(session);
@@ -662,6 +663,7 @@
 		g_signal_emit(media, purple_media_signals[STATE_CHANGED],
 				0, PURPLE_MEDIA_STATE_END,
 				NULL, participant);
+printf("free session 3\n");
 
 		if (link != NULL) {
 			g_free(link->data);
@@ -678,6 +680,7 @@
 		g_signal_emit(media, purple_media_signals[STATE_CHANGED],
 				0, PURPLE_MEDIA_STATE_END,
 				NULL, NULL);
+printf("free session 4\n");
 		g_object_unref(media);
 		return;
 	}
diff -r fc7b9e4d2d03 libpurple/protocols/jabber/jingle/jingle.c
--- a/libpurple/protocols/jabber/jingle/jingle.c	Tue Nov 26 00:30:54 2019 -0500
+++ b/libpurple/protocols/jabber/jingle/jingle.c	Thu Dec 12 09:33:37 2019 -0500
@@ -251,7 +251,7 @@
 jingle_handle_session_terminate(JingleSession *session, PurpleXmlNode *jingle)
 {
 	jabber_iq_send(jingle_session_create_ack(session, jingle));
-
+printf("here we receive terminate, remote hangup.\n");
 	jingle_session_handle_action(session, jingle,
 			JINGLE_SESSION_TERMINATE);
 	/* display reason? */
diff -r fc7b9e4d2d03 libpurple/protocols/jabber/jingle/rtp.c
--- a/libpurple/protocols/jabber/jingle/rtp.c	Tue Nov 26 00:30:54 2019 -0500
+++ b/libpurple/protocols/jabber/jingle/rtp.c	Thu Dec 12 09:33:37 2019 -0500
@@ -205,7 +205,7 @@
 		jabber_iq_send(jingle_session_terminate_packet(
 				session, type == PURPLE_MEDIA_INFO_HANGUP ?
 				"success" : "decline"));
-
+printf("here we hangup, send out terminate.\n");
 		g_signal_handlers_disconnect_by_func(G_OBJECT(media),
 				G_CALLBACK(jingle_rtp_state_changed_cb),
 				session);
diff -r fc7b9e4d2d03 meson.build
--- a/meson.build	Tue Nov 26 00:30:54 2019 -0500
+++ b/meson.build	Thu Dec 12 09:33:37 2019 -0500
@@ -136,8 +136,8 @@
 		 'wpurple_lib_dir("purple-@0@")'.format(purple_major_version))
 	conf.set('PIDGIN_LIBDIR',
 		 'wpurple_lib_dir("pidgin-@0@")'.format(purple_major_version))
-	conf.set('FINCH_LIBDIR',
-		 'wpurple_lib_dir("finch-@0@")'.format(purple_major_version))
+#	conf.set('FINCH_LIBDIR',
+#		 'wpurple_lib_dir("finch-@0@")'.format(purple_major_version))
 
 	conf.set('PURPLE_DATADIR', 'wpurple_data_dir()')
 	conf.set('PURPLE_SYSCONFDIR', 'wpurple_sysconf_dir()')
@@ -155,9 +155,9 @@
 	conf.set_quoted('PIDGIN_LIBDIR',
 	                join_paths(common_libdir,
 	                           'pidgin-@0@'.format(purple_major_version)))
-	conf.set_quoted('FINCH_LIBDIR',
-	                join_paths(common_libdir,
-	                           'finch-@0@'.format(purple_major_version)))
+#	conf.set_quoted('FINCH_LIBDIR',
+#	                join_paths(common_libdir,
+#	                           'finch-@0@'.format(purple_major_version)))
 endif
 
 abslibdir = join_paths(get_option('prefix'), get_option('libdir'))
@@ -165,8 +165,8 @@
 conf.set_quoted('PURPLE_PLUGINDIR', PURPLE_PLUGINDIR)
 PIDGIN_PLUGINDIR = join_paths(abslibdir, 'pidgin-@0@'.format(purple_major_version))
 conf.set_quoted('PIDGIN_PLUGINDIR', PIDGIN_PLUGINDIR)
-FINCH_PLUGINDIR = join_paths(abslibdir, 'finch-@0@'.format(purple_major_version))
-conf.set_quoted('FINCH_PLUGINDIR', FINCH_PLUGINDIR)
+#FINCH_PLUGINDIR = join_paths(abslibdir, 'finch-@0@'.format(purple_major_version))
+#conf.set_quoted('FINCH_PLUGINDIR', FINCH_PLUGINDIR)
 
 # Check for inet_aton
 if not IS_WIN32
@@ -766,7 +766,7 @@
 INSTALL_PIXMAPS = get_option('pixmaps-install')
 
 # check for gtk-doc
-ENABLE_DOC = get_option('doc')
+#ENABLE_DOC = get_option('doc')
 
 enable_debug = get_option('console-logging')
 conf.set('DEBUG', enable_debug)
@@ -776,9 +776,9 @@
 
 subdir('libpurple')
 subdir('share/sounds')
-subdir('finch')
+#subdir('finch')
 subdir('pidgin')
-subdir('doc')
+#subdir('doc')
 
 configure_file(output : 'config.h',
     configuration : conf)
@@ -788,7 +788,7 @@
 
 message('')
 message('Build GTK+ UI................. : ' + get_option('gtkui').to_string())
-message('Build console UI.............. : ' + enable_consoleui.to_string())
+#message('Build console UI.............. : ' + enable_consoleui.to_string())
 message('Build for X11................. : ' + x11.found().to_string())
 message('')
 message('Enable Gestures............... : ' + enable_gestures.to_string())
@@ -816,7 +816,7 @@
 message('Enable Introspection...........: ' + enable_introspection.to_string())
 message('')
 message('Print debugging messages...... : ' + enable_debug.to_string())
-message('Generate documentation........ : ' + ENABLE_DOC.to_string())
+#message('Generate documentation........ : ' + ENABLE_DOC.to_string())
 message('')
 bindir = join_paths(get_option('prefix'), get_option('bindir'))
 message('Pidgin will be installed in @0@.'.format(bindir))
diff -r fc7b9e4d2d03 pidgin/gtkblist.c
--- a/pidgin/gtkblist.c	Tue Nov 26 00:30:54 2019 -0500
+++ b/pidgin/gtkblist.c	Thu Dec 12 09:33:37 2019 -0500
@@ -1514,6 +1514,7 @@
 				PIDGIN_STOCK_TOOLBAR_AUDIO_CALL,
 				G_CALLBACK(gtk_blist_menu_audio_call_cb), buddy);
 		}
+printf("video call from main buddy list's right click menu.");
 		if (caps & PURPLE_MEDIA_CAPS_AUDIO_VIDEO) {
 			pidgin_new_menu_item(menu, _("Audio/_Video Call"),
 				PIDGIN_STOCK_TOOLBAR_VIDEO_CALL,
diff -r fc7b9e4d2d03 pidgin/gtkconv.c
--- a/pidgin/gtkconv.c	Tue Nov 26 00:30:54 2019 -0500
+++ b/pidgin/gtkconv.c	Thu Dec 12 09:33:37 2019 -0500
@@ -908,7 +908,7 @@
 	PidginConvWindow *win = (PidginConvWindow *)data;
 	PurpleConversation *conv = pidgin_conv_window_get_active_conversation(win);
 	PurpleAccount *account = purple_conversation_get_account(conv);
-
+printf("local start video call from conversation windows menu\n");
 	purple_protocol_initiate_media(account,
 			purple_conversation_get_name(conv),
 			action == win->menu->audio_call ? PURPLE_MEDIA_AUDIO :
diff -r fc7b9e4d2d03 pidgin/gtkmedia.c
--- a/pidgin/gtkmedia.c	Tue Nov 26 00:30:54 2019 -0500
+++ b/pidgin/gtkmedia.c	Thu Dec 12 09:33:37 2019 -0500
@@ -320,7 +320,7 @@
 			G_N_ELEMENTS(media_action_entries), media);
 
 	vbox = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
-	gtk_container_add(GTK_CONTAINER(media), vbox);
+	gtk_container_add(GTK_CONTAINER(media), vbox); // add to container
 
 	media->priv->statusbar = gtk_statusbar_new();
 	gtk_box_pack_end(GTK_BOX(vbox), media->priv->statusbar,
@@ -329,7 +329,7 @@
 			0, _("Calling..."));
 	gtk_widget_show(media->priv->statusbar);
 
-	media->priv->menubar = setup_menubar(media);
+	media->priv->menubar = setup_menubar(media); // create with builder
 	gtk_box_pack_start(GTK_BOX(vbox), media->priv->menubar,
 			FALSE, TRUE, 0);
 
@@ -339,7 +339,7 @@
 	gtk_box_pack_start(GTK_BOX(vbox), media->priv->display,
 			TRUE, TRUE, PIDGIN_HIG_BOX_SPACE);
 	gtk_widget_show(vbox);
-
+printf("here we display the video call window (vbox).\n");
 	g_signal_connect(G_OBJECT(media), "delete-event",
 			G_CALLBACK(pidgin_media_delete_event_cb), media);
 
@@ -641,11 +641,22 @@
 
 	gtkmedia->priv->request_type = PURPLE_MEDIA_NONE;
 	if (!purple_media_accepted(gtkmedia->priv->media, NULL, NULL)) {
+	    // emit "incoming-call"
+	    //purple_signal_emit(purple_xfers_get_handle(), "file-recv-request", xfer);
+	    //purple_signal_emit(purple_xfers_get_handle(), "incoming-call", gtkmedia->priv->media);
+	   
+	    // poll status if accepted
+	    // always accept for now
+	    if (1) {
+	      pidgin_media_accept_cb(gtkmedia->priv->media, 0);
+	    } else {
+	    //else -> ask
 		purple_request_accept_cancel(gtkmedia, _("Incoming Call"),
 			message, NULL, PURPLE_DEFAULT_ACTION_NONE,
 			purple_request_cpar_from_account(account),
 			gtkmedia->priv->media, pidgin_media_accept_cb,
 			pidgin_media_reject_cb);
+       }
 	}
 	pidgin_media_emit_message(gtkmedia, message);
 	g_free(message);
@@ -1048,12 +1059,13 @@
 	purple_debug_info("gtkmedia", "state: %d sid: %s name: %s\n",
 			state, sid ? sid : "(null)", name ? name : "(null)");
 	if (state == PURPLE_MEDIA_STATE_END) {
+printf("gtkmedia free session sid:%s, name:%s\n", sid, name);
 		if (sid != NULL && name != NULL) {
 			pidgin_media_remove_widget(gtkmedia, sid, name);
 		} else if (sid == NULL && name == NULL) {
 			pidgin_media_emit_message(gtkmedia,
 					_("The call has been terminated."));
-			gtk_widget_destroy(GTK_WIDGET(gtkmedia));
+			gtk_widget_destroy(GTK_WIDGET(gtkmedia)); // should call this if remote hangup
 		}
 	} else if (state == PURPLE_MEDIA_STATE_NEW &&
 			sid != NULL && name != NULL) {
diff -r fc7b9e4d2d03 pidgin/meson.build
--- a/pidgin/meson.build	Tue Nov 26 00:30:54 2019 -0500
+++ b/pidgin/meson.build	Thu Dec 12 09:33:37 2019 -0500
@@ -229,6 +229,7 @@
 		gnome.generate_gir(libpidgin,
 			sources : introspection_sources,
 			includes : ['GObject-2.0', 'Gtk-3.0', libpurple_gir[0], talkatu_gir, gplugin_gir, gplugin_gtk_gir],
+#			includes : ['GObject-2.0', 'Gtk-3.0', libpurple_gir[0], gplugin_gir, gplugin_gtk_gir],
 			namespace : 'Pidgin',
 			symbol_prefix : 'pidgin',
 			identifier_prefix : 'Pidgin',
